File: ios-parent/ScreenTimeParent/Shared/Networking/FamilyRepository.swift
// FamilyRepository.swift
// Screen Time Parent - Family data access layer

import Foundation
import Supabase

protocol FamilyRepositoryProtocol {
    func createFamily(name: String, createdBy: UUID) async throws -> Family
    func getFamily(id: UUID) async throws -> Family
    func getFamilyByInviteCode(_ code: String) async throws -> Family
    func getUserFamilies() async throws -> [FamilyOverview]
    func updateFamily(id: UUID, name: String) async throws -> Family
    func getFamilyMembers(familyId: UUID) async throws -> [FamilyMember]
    func addFamilyMember(request: JoinFamilyRequest) async throws -> FamilyMember
    func removeFamilyMember(id: UUID) async throws
    func getDevices(familyId: UUID) async throws -> [Device]
}

class FamilyRepository: FamilyRepositoryProtocol {
    private let client = SupabaseClientManager.shared.client

    func createFamily(name: String, createdBy: UUID) async throws -> Family {
        return try await RetryManager.shared.execute(operation: "FamilyRepository.createFamily") {
            let request = CreateFamilyRequest(name: name, createdBy: createdBy)
            return try await self.client.database
                .from("families")
                .insert(request)
                .select()
                .single()
                .execute()
                .value
        }
    }

    func getFamily(id: UUID) async throws -> Family {
        return try await RetryManager.shared.execute(operation: "FamilyRepository.getFamily") {
            try await self.client.database
                .from("families")
                .select()
                .eq("id", value: id.uuidString)
                .single()
                .execute()
                .value
        }
    }

    func getFamilyByInviteCode(_ code: String) async throws -> Family {
        return try await RetryManager.shared.execute(operation: "FamilyRepository.getFamilyByInviteCode") {
            try await self.client.database
                .from("families")
                .select()
                .eq("invite_code", value: code)
                .single()
                .execute()
                .value
        }
    }

    func getUserFamilies() async throws -> [FamilyOverview] {
        return try await RetryManager.shared.execute(operation: "FamilyRepository.getUserFamilies") {
            try await self.client.database
                .from("family_overview")
                .select()
                .execute()
                .value
        }
    }

    func updateFamily(id: UUID, name: String) async throws -> Family {
        return try await RetryManager.shared.execute(operation: "FamilyRepository.updateFamily") {
            try await self.client.database
                .from("families")
                .update(["name": name])
                .eq("id", value: id.uuidString)
                .select()
                .single()
                .execute()
                .value
        }
    }

    func getFamilyMembers(familyId: UUID) async throws -> [FamilyMember] {
        return try await RetryManager.shared.execute(operation: "FamilyRepository.getFamilyMembers") {
            try await self.client.database
                .from("family_members")
                .select("*, user_profiles(*)")
                .eq("family_id", value: familyId.uuidString)
                .execute()
                .value
        }
    }

    func addFamilyMember(request: JoinFamilyRequest) async throws -> FamilyMember {
        return try await RetryManager.shared.execute(operation: "FamilyRepository.addFamilyMember") {
            try await self.client.database
                .from("family_members")
                .insert(request)
                .select()
                .single()
                .execute()
                .value
        }
    }

    func removeFamilyMember(id: UUID) async throws {
        try await RetryManager.shared.execute(operation: "FamilyRepository.removeFamilyMember") {
            try await self.client.database
                .from("family_members")
                .delete()
                .eq("id", value: id.uuidString)
                .execute()
        }
    }

    func getDevices(familyId: UUID) async throws -> [Device] {
        return try await RetryManager.shared.execute(operation: "FamilyRepository.getDevices") {
            try await self.client.database
                .from("devices")
                .select()
                .eq("family_id", value: familyId.uuidString)
                .eq("is_active", value: true)
                .execute()
                .value
        }
    }
}

File: android-parent/app/src/main/java/com/screentimeparent/data/repository/FamilyRepository.kt
package com.screentimeparent.data.repository

import com.screentimeparent.data.models.*
import com.screentimeparent.data.remote.SupabaseClientProvider
import io.github.jan.supabase.auth.auth
import io.github.jan.supabase.postgrest.postgrest
import io.github.jan.supabase.postgrest.query.Columns
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import java.util.UUID
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class FamilyRepository @Inject constructor() {

    private val supabase = SupabaseClientProvider.client

    suspend fun createFamily(name: String): Result<Family> {
        return try {
            val userId = supabase.auth.currentUserOrNull()?.id
                ?: return Result.failure(Exception("User not authenticated"))

            val inviteCode = generateInviteCode()
            val familyId = UUID.randomUUID().toString()

            val family = Family(
                id = familyId,
                name = name,
                inviteCode = inviteCode,
                createdBy = userId
            )

            supabase.postgrest["families"].insert(family)

            // Add creator as admin
            val member = FamilyMember(
                id = UUID.randomUUID().toString(),
                familyId = familyId,
                userId = userId,
                role = FamilyMemberRole.ADMIN
            )
            supabase.postgrest["family_members"].insert(member)

            Result.success(family)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    suspend fun getCurrentFamily(): Family? {
        val userId = supabase.auth.currentUserOrNull()?.id ?: return null

        return try {
            val membership = supabase.postgrest["family_members"]
                .select {
                    filter {
                        eq("user_id", userId)
                    }
                }
                .decodeSingleOrNull<FamilyMember>()

            if (membership != null) {
                supabase.postgrest["families"]
                    .select {
                        filter {
                            eq("id", membership.familyId)
                        }
                    }
                    .decodeSingle<Family>()
            } else {
                null
            }
        } catch (e: Exception) {
            null
        }
    }

    suspend fun getCurrentFamilyOverview(): FamilyOverview? {
        val family = getCurrentFamily() ?: return null

        return try {
            val members = supabase.postgrest["family_members"]
                .select {
                    filter {
                        eq("family_id", family.id)
                    }
                }
                .decodeList<FamilyMember>()

            val childCount = members.count { it.role == FamilyMemberRole.CHILD }
            val parentCount = members.count { it.role != FamilyMemberRole.CHILD }

            val deviceCount = supabase.postgrest["devices"]
                .select {
                    filter {
                        eq("family_id", family.id)
                    }
                }
                .decodeList<Device>().size

            FamilyOverview(
                id = family.id,
                name = family.name,
                inviteCode = family.inviteCode,
                childCount = childCount,
                parentCount = parentCount,
                deviceCount = deviceCount
            )
        } catch (e: Exception) {
            null
        }
    }

    suspend fun getFamilyMembers(familyId: String): List<FamilyMember> {
        return try {
            supabase.postgrest["family_members"]
                .select(Columns.raw("*, user_profile:user_profiles(*)")) {
                    filter {
                        eq("family_id", familyId)
                    }
                }
                .decodeList<FamilyMember>()
        } catch (e: Exception) {
            emptyList()
        }
    }

    suspend fun getChildStatuses(familyId: String): List<ChildStatus> {
        return try {
            val members = getFamilyMembers(familyId)
            val children = members.filter { it.role == FamilyMemberRole.CHILD }

            children.mapNotNull { member ->
                val profile = member.userProfile ?: return@mapNotNull null

                // Get device info
                val device = supabase.postgrest["devices"]
                    .select {
                        filter {
                            eq("child_id", member.userId)
                        }
                    }
                    .decodeSingleOrNull<Device>()

                // Get today's earned time
                val today = java.time.LocalDate.now().toString()
                val earnedTime = supabase.postgrest["earned_time"]
                    .select {
                        filter {
                            eq("child_id", member.userId)
                            eq("date", today)
                        }
                    }
                    .decodeSingleOrNull<EarnedTime>()

                ChildStatus(
                    childId = member.userId,
                    childName = member.nickname ?: profile.fullName,
                    deviceName = device?.deviceName,
                    educationalMinutes = earnedTime?.educationalMinutes ?: 0,
                    requiredEducationalMinutes = earnedTime?.requiredEducationalMinutes ?: 30,
                    recreationalMinutesUsed = earnedTime?.recreationalMinutesUsed ?: 0,
                    recreationalMinutesAvailable = earnedTime?.recreationalMinutesAvailable,
                    requirementMet = earnedTime?.requirementMet ?: false,
                    lastSync = device?.lastSync,
                    isOnline = device?.isActive ?: false
                )
            }
        } catch (e: Exception) {
            emptyList()
        }
    }

    suspend fun joinFamily(inviteCode: String): Result<Family> {
        return try {
            val userId = supabase.auth.currentUserOrNull()?.id
                ?: return Result.failure(Exception("User not authenticated"))

            val family = supabase.postgrest["families"]
                .select {
                    filter {
                        eq("invite_code", inviteCode.uppercase())
                    }
                }
                .decodeSingleOrNull<Family>()
                ?: return Result.failure(Exception("Invalid invite code"))

            // Check if already a member
            val existingMember = supabase.postgrest["family_members"]
                .select {
                    filter {
                        eq("family_id", family.id)
                        eq("user_id", userId)
                    }
                }
                .decodeSingleOrNull<FamilyMember>()

            if (existingMember != null) {
                return Result.failure(Exception("Already a member of this family"))
            }

            // Add as parent
            val member = FamilyMember(
                id = UUID.randomUUID().toString(),
                familyId = family.id,
                userId = userId,
                role = FamilyMemberRole.PARENT
            )
            supabase.postgrest["family_members"].insert(member)

            Result.success(family)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    private fun generateInviteCode(): String {
        val chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
        return (1..6).map { chars.random() }.joinToString("")
    }

    fun observeFamilyUpdates(familyId: String): Flow<Family?> = flow {
        // Polling-based updates
        while (true) {
            try {
                val family = supabase.postgrest["families"]
                    .select {
                        filter {
                            eq("id", familyId)
                        }
                    }
                    .decodeSingleOrNull<Family>()
                emit(family)
            } catch (e: Exception) {
                emit(null)
            }
            kotlinx.coroutines.delay(60000) // Poll every 60 seconds
        }
    }
}
